osquery/database/plugins/rocksdb.cpp

```diff
@@ -114,6 +120,41 @@ class RocksDBDatabasePlugin : public DatabasePlugin {
 /// Backing-storage provider for osquery internal/core.
 REGISTER_INTERNAL(RocksDBDatabasePlugin, "database", "rocksdb");
 
+class RocksDBGuard {
+ public:
+  RocksDBGuard() : my_tid_(platformGetTid()), locked_(false) {}
+
+  bool safeEnter() {
+    if (!owner_tid_is_valid_ || owner_tid_ != my_tid_) {
+      mu_.lock();
```

Comment:
I think I was too quick with my above comment. So, this won't protect against the rocksdb lock vs google lock deadlock situation:

Thread x: Google Log -> Takes Google Mutex -> calls BufferedLogSink::send() -> Calls Rocks DB Put -> attempts memtable (or some other internal) mutex

Thread y: RocksDB BG thread -> Rocks DB Flush -> Takes memtable/internal mutex -> Logging Callback -> Google Logging -> Attempts Google Lock.

Your recommended solution doesn't protect against this. In other words, there are two separate, but related, deadlock situations here - one caused by non-reentrancy of rocksdb API calls, and the other caused by inverted acquisition of locks in two separate modules (Google Logging and RocksDB Logging)

Incidentally, in case anyone is wondering why this deadlock doesn't manifest itself all the time, it's because it requires enough rapid calls to rocksdb put to trigger a flush in the calling thread, followed by just enough logs in Google Logging to force it to do a flush as well. We have not been able to trigger this situation even with aggressive query schedules.

However, a "flood" of FIM events - untar-ing 50 to 200 files into a monitored directory - causes a burst of events to be recorded, and it deadlocks rather easily.